🐛 Bug Report: Custom Entity Providers documentation incomplete

### 📜 Description

Referencing my issue #25404, from my work account, the [Custom Entity Providers](https://backstage.io/docs/features/software-catalog/external-integrations/#custom-entity-providers) doc is incomplete and needs to specify properly what you should do in the new Backend.

After following the guide and creating the `FrobsProvider` class in the old Backend you [install the provider](https://backstage.io/docs/features/software-catalog/external-integrations/#installing-the-provider) like so 
```
  const frobs = new FrobsProvider('production', env.reader);
  builder.addEntityProvider(frobs);

  const { processingEngine, router } = await builder.build();
  await processingEngine.start();

  await env.scheduler.scheduleTask({
    id: 'run_frobs_refresh',
    fn: async () => {
      await frobs.run();
    },
    frequency: { minutes: 30 },
    timeout: { minutes: 10 },
  });
```

while in the [new Backend](https://backstage.io/docs/features/software-catalog/external-integrations/#new-backend-system) it say to do it like so:
```
export const catalogModuleFrobsProvider = createBackendModule({
  pluginId: 'catalog',
  moduleId: 'frobs-provider',
  register(env) {
    env.registerInit({
      deps: {
        catalog: catalogProcessingExtensionPoint,
        reader: coreServices.urlReader,
      },
      async init({ catalog, reader }) {
        catalog.addEntityProvider(new FrobsProvider('dev', reader));
      },
    });
  },
});
```

As you can see, the whole scheduler is missing but especially this bit of code inside the scheduler is missing `await frobs.run();` in the new Backend and there is no mention of how to add it

### 👍 Expected behavior

The new Backend documentation should mention how to trigger the `run` function after installation, and to be completed, it would be nice to add also how you can add a scheduler.

### 👎 Actual Behavior with Screenshots

After following the documentation you end up with the new entity provider not working and without knowing the real reason why only the logger inside this function is triggered.

```
  async connect(connection: EntityProviderConnection): Promise<void> {
    this.logger.info('Establishing connection.');
    this.connection = connection;
  }
```

### 👟 Reproduction steps

- migrate the catalog to the new Backend
- add a new entity provider following the documentation above

### 📃 Provide the context for the Bug.

_No response_

### 🖥️ Your Environment

_No response_

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

None

hey @aramissennyeydd @freben today I externalised the user provider in a BE module and `yarn tsc` don't pass anymore.

if I follow the documentation at some point it says to use this command `yarn new --select backend-module --option id=catalog` to create a proper module, it creates everything and adds the module to the backend package.json which works perfectly but then once I commit the pipeline fail on the `tsc` part `Cannot find module '@internal/backstage-plugin-catalog-backend-module-okta-user-provider' or its corresponding type declarations.`

--- a/docs/features/software-catalog/external-integrations.md
+++ b/docs/features/software-catalog/external-integrations.md
@@ -310,7 +310,7 @@ export const catalogModuleFrobsProvider = createBackendModule({
         scheduler: coreServices.scheduler,
         /* highlight-add-end */
       },
-      async init({ catalog, reader, scheduler, config }) {
+      async init({ catalog, reader, scheduler }) {
         const taskRunner = scheduler.createScheduledTaskRunner({
           frequency: { minutes: 30 },
           timeout: { minutes: 10 },

--- a/docs/features/software-catalog/external-integrations.md
+++ b/docs/features/software-catalog/external-integrations.md
@@ -84,6 +84,7 @@ import {
   EntityProvider,
   EntityProviderConnection,
 } from '@backstage/plugin-catalog-node';
+import { SchedulerServiceTaskRunner } from '@backstage/backend-plugin-api';
 
 /**
  * Provides entities from fictional frobs service.
@@ -92,11 +93,17 @@ export class FrobsProvider implements EntityProvider {
   private readonly env: string;
   private readonly reader: UrlReader;
   private connection?: EntityProviderConnection;
+  private taskRunner: SchedulerServiceTaskRunner;
 
   /** [1] */
-  constructor(env: string, reader: UrlReader) {
+  constructor(
+    env: string,
+    reader: UrlReader,
+    taskRunner: SchedulerServiceTaskRunner,
+  ) {
     this.env = env;
     this.reader = reader;
+    this.taskRunner = taskRunner;
   }
 
   /** [2] */
@@ -107,6 +114,12 @@ export class FrobsProvider implements EntityProvider {
   /** [3] */
   async connect(connection: EntityProviderConnection): Promise<void> {
     this.connection = connection;
+    this.taskRunner.run({
+      id: this.getProviderName(),
+      fn: async () => {
+        await this.run();
+      },
+    });
   }
 
   /** [4] */
@@ -248,23 +261,16 @@ export default async function createPlugin(
 ): Promise<Router> {
   const builder = CatalogBuilder.create(env);
   /* highlight-add-start */
-  const frobs = new FrobsProvider('production', env.reader);
+  const taskRunner = env.scheduler.createScheduledTaskRunner({
+    frequency: { minutes: 30 },
+    timeout: { minutes: 10 },
+  });
+  const frobs = new FrobsProvider('production', env.reader, taskRunner);
   builder.addEntityProvider(frobs);
   /* highlight-add-end */
 
   const { processingEngine, router } = await builder.build();
   await processingEngine.start();
-
-  /* highlight-add-start */
-  await env.scheduler.scheduleTask({
-    id: 'run_frobs_refresh',
-    fn: async () => {
-      await frobs.run();
-    },
-    frequency: { minutes: 30 },
-    timeout: { minutes: 10 },
-  });
-  /* highlight-add-end */
 
   // ..
 }
@@ -300,19 +306,17 @@ export const catalogModuleFrobsProvider = createBackendModule({
       deps: {
         catalog: catalogProcessingExtensionPoint,
         reader: coreServices.urlReader,
+        /* highlight-add-start */
         scheduler: coreServices.scheduler,
+        /* highlight-add-end */
       },
-      async init({ catalog, reader, scheduler }) {
-        const frobs = new FrobsProvider('dev', reader);
-        catalog.addEntityProvider(frobs);
-        await scheduler.scheduleTask({
-          id: 'run_frobs_refresh',
-          fn: async () => {
-            await frobs.run();
-          },
+      async init({ catalog, reader, scheduler, config }) {
+        const taskRunner = scheduler.createScheduledTaskRunner({
           frequency: { minutes: 30 },
           timeout: { minutes: 10 },
         });
+        const frobs = new FrobsProvider('dev', reader, taskRunner);
+        catalog.addEntityProvider(frobs);
       },
     });
   },
@@ -326,6 +330,66 @@ backend.add(catalogModuleFrobsProvider);
 // Other plugins ...
 
 backend.start();
+```
+
+#### Follow-up: Config Defined Schedule
+
+If you want to go a step further and increase the configurability of your new `FrobsProvider`, you can define the schedule that the task runs at in `app-config.yaml` instead of requiring code changes to adjust.
+
+```yaml
+catalog:
+  providers:
+    frobs-provider:
+      schedule:
+        initialDelay: { seconds: 30 }
+        frequency: { hours: 1 }
+        timeout: { minutes: 50 }
+```
+
+This approach will also allow you to customize the schedule per environment
+
+#### New Backend
+
+```ts
+import {
+  SchedulerServiceTaskScheduleDefinition,
+  /* highlight-add-start */
+  readSchedulerServiceTaskScheduleDefinitionFromConfig,
+  /* highlight-add-end */
+} from '@backstage/backend-plugin-api';
+
+export const catalogModuleFrobsProvider = createBackendModule({
+  pluginId: 'catalog',
+  moduleId: 'frobs-provider',
+  register(env) {
+    env.registerInit({
+      deps: {
+        // ... other deps
+        /* highlight-add-start */
+        rootConfig: coreServices.rootConfig,
+        /* highlight-add-end */
+      },
+      async init({ catalog, reader, scheduler, rootConfig }) {
+        /* highlight-add-start */
+        const config = rootConfig.getConfig('catalog.providers.frobs-provider'); // Generally, catalog config goes under catalog.providers.pluginId
+        // Add a default schedule if you don't define one in config.
+        const schedule = config.has('schedule')
+          ? readSchedulerServiceTaskScheduleDefinitionFromConfig(
+              config.getConfig('schedule'),
+            )
+          : {
+              frequency: { minutes: 30 },
+              timeout: { minutes: 10 },
+            };
+        const taskRunner: SchedulerServiceTaskRunner =
+          scheduler.createScheduledTaskRunner(schedule);
+        /* highlight-add-end */
+
+        // rest of your code
+      },
+    });
+  },
+});
 ```
 
 ### Example User Entity Provider

--- a/docs/features/software-catalog/external-integrations.md
+++ b/docs/features/software-catalog/external-integrations.md
@@ -300,9 +300,19 @@ export const catalogModuleFrobsProvider = createBackendModule({
       deps: {
         catalog: catalogProcessingExtensionPoint,
         reader: coreServices.urlReader,
+        scheduler: coreServices.scheduler,
       },
-      async init({ catalog, reader }) {
-        catalog.addEntityProvider(new FrobsProvider('dev', reader));
+      async init({ catalog, reader, scheduler }) {
+        const frobs = new FrobsProvider('dev', reader);
+        catalog.addEntityProvider(frobs);
+        await scheduler.scheduleTask({
+          id: 'run_frobs_refresh',
+          fn: async () => {
+            await frobs.run();
+          },
+          frequency: { minutes: 30 },
+          timeout: { minutes: 10 },
+        });
       },
     });
   },

--- a/docs/features/software-catalog/external-integrations.md
+++ b/docs/features/software-catalog/external-integrations.md
@@ -336,7 +336,7 @@ backend.start();
 
 If you want to go a step further and increase the configurability of your new `FrobsProvider`, you can define the schedule that the task runs at in `app-config.yaml` instead of requiring code changes to adjust.
 
-```yaml
+```yaml title="app-config.yaml"
 catalog:
   providers:
     frobs-provider:
@@ -350,7 +350,7 @@ This approach will also allow you to customize the schedule per environment
 
 #### New Backend
 
-```ts
+```ts title="packages/backend/src/index.ts"
 import {
   SchedulerServiceTaskScheduleDefinition,
   /* highlight-add-start */
@@ -390,6 +390,38 @@ export const catalogModuleFrobsProvider = createBackendModule({
     });
   },
 });
+```
+
+#### Old Backend
+
+```ts title="packages/backend/src/plugins/catalog.ts"
+/* highlight-add-next-line */
+import { FrobsProvider } from '../path/to/class';
+import {
+  /* highlight-add-start */
+  readSchedulerServiceTaskScheduleDefinitionFromConfig,
+  /* highlight-add-end */
+} from '@backstage/backend-plugin-api';
+
+export default async function createPlugin(
+  env: PluginEnvironment,
+): Promise<Router> {
+  /* highlight-add-start */
+  const config = env.config.getConfig('catalog.providers.frobs-provider'); // Generally, catalog config goes under catalog.providers.pluginId
+  // Add a default schedule if you don't define one in config.
+  const schedule = config.has('schedule')
+    ? readSchedulerServiceTaskScheduleDefinitionFromConfig(
+        config.getConfig('schedule'),
+      )
+    : {
+        frequency: { minutes: 30 },
+        timeout: { minutes: 10 },
+      };
+  const taskRunner = env.scheduler.createScheduledTaskRunner(schedule);
+  /* highlight-add-end */
+
+  // ..
+}
 ```
 
 ### Example User Entity Provider

